src/esp/assets/GenericInstanceMeshData.cpp

```diff
@@ -94,33 +105,183 @@ Cr::Containers::Optional<InstancePlyData> parsePly(
+      // of objectIDs
+      data.objPartitionsFromSSD = true;
+
+      // from SSD's objects_ array - this is the number of defined semantic
+      // descriptors corresponding to object instances.
+      // all this tapdancing because of the design of the Semantic Scene.
+      const auto& ssdObjs = semanticScene->objects();
+      std::size_t numSSDObjs = ssdObjs.size();
+
+      // map of colors as ints as key, where value is objectID and regionID from
+      // ssd
+      std::unordered_map<uint32_t, std::pair<int, int>>
+          tmpColorMapToSSDidAndRegionIndex;
+      tmpColorMapToSSDidAndRegionIndex.reserve(numSSDObjs);
+      // lambda function to convert a 3-vec representation of a color into an
+      // int
+      auto colorAsInt = [](const Mn::Vector3ub& color) -> uint32_t {
+        return (((color[0] << 16) + color[1]) << 16) + color[2];
+      };
+
+      // build maps of color ints to semantic IDs and color ints to region ids
+      // in SSD
+      // find max regions present, so we can have an "unassigned"/"unknown"
+      // region to move all verts whose region is
+
+      int maxRegion = -1;
+      for (int i = 0; i < numSSDObjs; ++i) {
+        const auto& ssdObj =
+            static_cast<scene::HM3DObjectInstance&>(*ssdObjs[i]);
+        const uint32_t colorInt = colorAsInt(ssdObj.getColor());
+        int regionIDX =
+            static_cast<scene::HM3DSemanticRegion&>(*ssdObj.region())
+                .getIndex();
+        tmpColorMapToSSDidAndRegionIndex[colorInt] = {ssdObj.getSemanticID(),
+                                                      regionIDX};
+        maxRegion = Mn::Math::max(regionIDX, maxRegion);
+      }
+
+      // increment maxRegion to use for unknown regions whose colors are not
+      // mapped
+      ++maxRegion;
+
+      // rebuild objectIDs vector (objectID per vertex) based on
+      // per vertex colors mapped
+      // 1st semantic ID for colors not found in SSD
+      std::size_t nonSSDObjID = tmpColorMapToSSDidAndRegionIndex.size();
+
+      // map regionIDs to affected verts using semantic color provided in
+      // SemanticScene, as specified in SSD file.
+      data.partitionIds.resize(meshData->vertexCount());
+      // map semantic IDs corresponding to colors from SSD file to appropriate
+      // verts (via index)
+      data.objectIds.resize(meshData->vertexCount());
+      // temporary holding structure to hold any non-SSD vert colors, so that
+      // the nonSSDObjID for new colors can be incremented appropriately
+      // not using set to avoid extra include
+      std::unordered_map<uint32_t, int> nonSSDVertColors;
+
+      // temporary map keyed by semantic ID and value being actual color for
+      // that ID. We need this since we do not know how many, if any, unique
+      // colors that do not have Semantic Mappings exist on the mesh
+      std::unordered_map<int, Mn::Vector3ub> tmpDestSSDColorMap;
```

Comment:
Instead of this, why not `resize()` the `colorMapToUse` to `id + 1` that you're about to insert, and then `colorMapToUse[id] = stuff` directly? You're copying this to a linear array after anyway so this won't save anything.

Reply:
id is different for each vert - can get smaller, so don't want to resize colorMapToUse smaller than it was for some previous vert.

Reply:
Err, I meant this:

```cpp
if(colorMapToUse.size() <= id) colorMapToUse.resize(id + 1);
colorMapToUse[id] = stuff;
```

Using a hashmap for something that's basically contiguous indexing doesn't seem right to me :)

Reply:
Except the id does not necessarily have to be contiguous. It is provided in the semantic lexicon text document, and so long as it is unique, it can be anything.